blog/2023-11-01-feature-flags-sdks-architectures.md

```diff
@@ -0,0 +1,159 @@
+  microservice
+  initialization is handled by not evaluating to the expected feature flag value.
+* First requests are not in cache, thus slower and causing network traffic.
+* Not every request can be efficiently cached, as some evaluations can be dynamic according to context with different
+  values.
+* When there is no background task for getting real-time configuration changes event and refreshing the cache,
+  values can take time to be updated.
+
+### Local evaluation
+
+At this approach, the Feature Flags configuration is saved locally at the SDK.  
+Feature flag evaluation is done on locally via the SDK and not affecting network traffic.
+With this approach, configuration can be fetched on initialization, and be refreshed periodically and/or when there is
+a configuration change, which can be triggered at the SDK via stream event from the Feature Flags service.
+some SDKs sync data via WebSockets or [SSE](https://en.wikipedia.org/wiki/Server-sent_events).
+When this approach is used, evaluation analytics metrics can be aggregated at the SDK side and published separately, as
+not every flag evaluation triggers a message to the Features Service / Proxy.
+
+### Local evaluation - example flow
+
+<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Local-evaluation.png').default} />
+
+#### Advantages
+
+* Functional advantage is that intermediate issue like network or feature flags service error is automatically handled,
+  as configuration is cached, and flags are evaluated locally.
+  If an issue like network error is ongoing for a long time period, SDK provider state can also be reflected as stale.
+* Another functional advantage, initialization can have a fallback mechanism to dealing with situations like temporary
+  network failure by starting from last received configuration or some other configuration, thus microservice can start
+  with last feature flag configuration.
+* Feature flags configuration can be cached on CDN.
+* Feature flag evaluation is done on locally via the SDK and not affecting network traffic, hence can be fast and
+  efficient.
+* Working without a relay proxy can still be efficient from SDK point of view, as flags are evaluated locally and
+  network traffic does not block common
+  flow.
+
+#### Disadvantages
+
+* When feature flags configuration is very large with many feature flags, network traffic and local cache sizes can be
+  increased.
+* Local evaluation should be implemented at the SDK, developed and maintained for multiple languages. Considering rules
+  like context properties based and percentage rollout strategies via hashing to buckets, it can be hard to keep 100%
+  feature parity across languages.
+
+#### A note on client-side SDKs for web browser / client app
+
+This approach is less suitable as is for client-side SDKs from security point of view, as full feature flags
+configuration and authorization key are exposed to end users.
```

Comment:
small grammatical changes, to read:
This approach is less suitable for client-side SDKs from a security point of view, as full feature flag configuration and authorization keys are exposed to end users.